syntax = "proto3";
package remote_torch;

message Reference {
    string identifier = 1;
    string description = 2;
}

message Chunk {
    bytes data = 1;
    string description = 2;
    bytes secret = 3;
}

message Empty {}

message TrainConfig {
    Reference model = 1;
    Reference dataset = 2;
    int32 batch_size = 3;
    int32 epochs = 4;
    string device = 5;
    string metric = 6;

    oneof privacy {
        Empty standard = 7;
        DpParameters differential_privacy = 8;
    }
    
    oneof optimizer {
        SGD sgd = 9;
        Adam adam = 10;
    }

    message DpParameters {
        float max_grad_norm = 1;
        float noise_multiplier = 2;
    }

    message SGD {
        float learning_rate = 1;
        float weight_decay = 2;
        float momentum = 3;
        float dampening = 4;
        bool nesterov = 5;
    }

    message Adam {
        float learning_rate = 1;
        float beta_1 = 2;
        float beta_2 = 3;
        float epsilon = 4;
        float weight_decay = 5;
        bool amsgrad = 6;
    }
}

message TestConfig {
    Reference model = 1;
    Reference dataset = 2;
    int32 batch_size = 3;
    string metric = 4;
}

message References {
    repeated Reference list = 1;
}  

message Accuracy {
    float value = 1;
}

message Devices {
    repeated string list = 1;
}

service RemoteTorch {
    rpc SendDataset (stream Chunk) returns (Reference) {}
    rpc SendModel (stream Chunk) returns (Reference) {}
    rpc FetchDataset (Reference) returns (stream Chunk) {}
    rpc FetchModule (Reference) returns (stream Chunk) {}
    rpc DeleteDataset (Reference) returns (Empty) {}
    rpc DeleteModule (Reference) returns (Empty) {}
    rpc AvailableModels(Empty) returns (References) {}
    rpc AvailableDatasets(Empty) returns (References) {}
    rpc AvailableDevices(Empty) returns (Devices) {}
    rpc Train (TrainConfig) returns (Empty) {}
    rpc Test (TestConfig) returns (Accuracy) {}
}
